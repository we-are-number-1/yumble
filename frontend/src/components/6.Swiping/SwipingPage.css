.TimeCounter {
  color: white;
  font-family: Comic Sans MS;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: 2.5em;
  margin-top: 0.5vw;
}

.btnColumn {
  display: flex;
  align-items: center;
  justify-content: center;
}

.YesOrNoButton {
  cursor: pointer;
  outline: none;
  border: 4px solid #000000;
  color: white;
  padding: 40px;
  text-align: center;
  font-size: 1.5em;
  margin: 4px 2px;
  border-radius: 50%;
}

.hide {
  visibility: hidden;
}

#YesButton {
  background-color: #6dc070; /* Green */
}

#YesButton:hover {
  background-color: #85c988;
}

#NoButton {
  background-color: #d34836; /* Green */
}

#NoButton:hover {
  background-color: #d66354;
}
