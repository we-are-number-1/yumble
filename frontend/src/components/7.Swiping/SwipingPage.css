
.TimeCounter {
    color: white;
    font-family: Comic Sans MS;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: 2.5vw;
    margin-top: 0.5vw;
  }
  
  .YesOrNoButton{
    cursor: pointer;
    position: absolute;
    outline: none;
    top: 20vw;
    border: 4px solid #000000;
    color: white;
    padding: 40px;
    text-align: center;
    font-size: 32px;
    margin: 4px 2px;
    border-radius: 50%;
  }
  
  #YesButton{
    background-color: #6dc070; /* Green */
    right: 15vw;
  }
  
  #YesButton:hover{
    background-color: #85c988;
  }
  
  #NoButton{
    background-color: #d34836; /* Green */
    left: 15vw
  }
  
  #NoButton:hover{
    background-color: #d66354;
  }
  
  