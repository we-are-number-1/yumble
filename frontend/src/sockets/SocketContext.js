import React, {useState} from 'react';
import {io} from 'socket.io-client';

const SocketContext = React.createContext();

/**
 * @param {*} param0
 * @return {*} SocketContext Provider
 */
function SocketContextProvider({children}) {
  const [socket, setSocket] = useState(() => {
    setSocket(io());
  });
  const [countdown, setCountdown] = useState(null);
  const [users, setUsers] = useState(null);
  const [preferences, setPreferences] = useState(null);

  const context = {
    socket,
    countdown,
    setCountdown,
    users,
    setUsers,
    preferences,
    setPreferences,
  };

  return (
    <SocketContext.Provider value={context}>
      {children}
    </SocketContext.Provider>
  );
}

export {
  SocketContext,
  SocketContextProvider,
};
