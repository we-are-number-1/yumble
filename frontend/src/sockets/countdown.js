/**
 * Function for processing game start countdown
 * @param {*} socket
 * @param {*} cb call back function
 */
export function countdown(socket, cb) {
  socket.on('countdown', ({count}) => {
    // Call back function with current count
    cb(count);
  });
}
