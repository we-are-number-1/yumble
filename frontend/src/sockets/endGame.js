/**
 * Function for when the game ends.
 * @param {*} socket
 * @param {*} cb call back function
 */
export function endGame(socket, cb) {
  socket.on('end_game', (data) => {
    socket.disconnect(true);
    cb(data);
  });
}
