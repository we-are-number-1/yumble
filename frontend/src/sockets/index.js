export {getSocketClient} from './socketClient';
export {newUser} from './newUser';
