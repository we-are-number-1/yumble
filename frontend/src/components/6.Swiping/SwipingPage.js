import React, { useEffect, useState, useContext, useRef } from 'react';
import { Redirect } from 'react-router-dom';

import { SocketContext } from './../../sockets/SocketContext';
import * as SocketEvents from './../../sockets';
import Help from '../Common/Help';
import '../Common/Help.css';
import MapModal from '../Common/MapModal';
import Icon from '../Common/MapsPinpoint';
import '../6.Swiping/SwipingPage.css';
import SwipeCard from '../Common/SwipeCard';
import { Container, Row, Col } from 'react-bootstrap';
import TinderCard from 'react-tinder-card'

/**
 * @param  {*} props
 * @return {*}
 */
function SwipingPage(props) {
  const socketContext = useContext(SocketContext);
  const [ButtonPopup, setButtonPopup] = useState(false);
  const [MapPopup, setMapPopup] = useState(false);
  const CardData = props.location.state[0];
  const [CardPass, setCardPass] = useState(null);
  const [time, setTime] = useState(socketContext.timer);
  const [vote, setVote] = useState(undefined);
  const [showVoteButtons, setShowVoteButtons] = useState(true);
  const [Data, setData] = useState(CardData[0]);
  const [redirect, setRedirect] = useState(false);
  const swiperRef = useRef()
  const newCard = useRef()

  useEffect(() => {
    document.title = 'Yes or No?';
    const card = CardData.shift();
    newCard.current = card;

    setData(card);
    SocketEvents.endGame(socketContext.socket, goNextPge);
    SocketEvents.nextRound(socketContext.socket, getNewCard);
    setCardPass(props.location.state[0].slice());
  }, []);

  useEffect(() => {
    const t = setTimeout(() => {
      if (time > 0) {
        setTime(time - 1);
      }
    }, 1000);

    return () => {
      clearTimeout(t);
    };
  }, [time]);

  const goNextPge = () => {
    setRedirect(true);
  };

  /**
   * Swipes card right when clicking yes
   */
  function clickedYes() {
    if(swiperRef.current) {
      swiperRef.current.swipe('right');
    }
  }

  /**
   * Swipes card left when clicking no
   */
  function clickedNo() {
    if(swiperRef.current) {
      swiperRef.current.swipe('left');
    }
  }
  
  /**
   * This is required as handleCardLeftScreen is called much later than 
   * when the swipe event is raised. 
   * @param {*} direction 
   */
  function handleSwipe(direction) {
    // hide vote buttons when swiped
    if(vote !== undefined) {
      return;
    }

    switch(direction) {
      case 'left':
        setShowVoteButtons(false);
        break;
      case 'right':
        SocketEvents.vote(socketContext.socket, socketContext.code, {
          name: Data.name,
          location: Data.location,
          coords: Data.coords,
          price: Data.price,
          rating: Data.rating,
          images: Data.images,
        });

        setShowVoteButtons(false);

        break;
    }
  }

  /**
   * Handle a swipe of a card to a specific direction.
   * If left, then vote is set to false
   * If right, then vote is set to right and send to server.
   * If already voted, then do nothing
   * @param {*} direction 
   */
  function handleCardLeftScreen(direction) {
    // If we recieve a new card during the time that card is still being swiped then we ignore the previous swipe
    if(newCard.current !== Data) {
      return;
    }

    switch(direction) {
      case 'left':
        setVote(false);
        break;
      case 'right':
        setVote(true);
    }
  }

  /**
   * @param {*} timer new restaurant details
   * @return {void}
   */
  function getNewCard() {
    setTime(socketContext.timer);
    try {
      setVote(undefined)
      setShowVoteButtons(true)
      CardData.shift();
      newCard.current = CardData[0];

      if (CardData[0] !== undefined) {
        setData(CardData[0]);
      }
    } catch (error) {}
  }

  let voteString = ""
  let overlayStyling = ""
  if (vote !== undefined){
    voteString = vote ? "Keen" : "Nope"
    overlayStyling = vote ? "CardOverlayText-Keen" : "CardOverlayText-Nope"
  }
  const buttonHideStyling = showVoteButtons ? "" : "hide";
  return (
    <>
      <h1 className='Title'> yumble</h1>
      <h1 className='TimeCounter'> Remaining time: {time}s</h1>
      <div className='MakeCentre'>
        <Container style={{ marginTop: '2em' }}>
          <Row lg={12} className='justify-content-md-center'>
            <Col
              xs={{ span: 6, order: 2 }}
              md={{ span: 2, order: 1 }}
              className='btnColumn'
            >
              <button
                className={'YesOrNoButton ' + buttonHideStyling}
                id='NoButton'
                onClick={clickedNo}
              >
                Nope!
              </button>
            </Col>
            <Col lg={6} xs={{ span: 12, order: 1 }} md={{ span: 8, order: 2 }}>
              <div className={"CardOverlayText " + overlayStyling}>Voted: {vote === undefined ? "" : voteString}!</div>
              {
                vote === undefined ? (
                  <TinderCard className="ActionableSwipeCard" key={Data.name} ref={swiperRef} onSwipe={handleSwipe} onCardLeftScreen={handleCardLeftScreen} preventSwipe={['up', 'down']}>
                    <SwipeCard data={Data} />
                  </TinderCard>
                ) : <SwipeCard vote={vote} data={Data} />
              }
            </Col>
            <Col
              xs={{ span: 6, order: 3 }}
              md={{ span: 2, order: 3 }}
              className='btnColumn'
            >
              <button
                className={"YesOrNoButton align-items-center " + buttonHideStyling}
                id='YesButton'
                onClick={clickedYes}
              >
                Keen!
              </button>
            </Col>
          </Row>
        </Container>

        <button
          onClick={() => setMapPopup(true)}
          className='BigBtn'
          id='GoogleMaps_btn'
        >
          View on Google Maps
          <Icon />
        </button>
        <MapModal
          trigger={MapPopup}
          setTrigger={setMapPopup}
          restaurantLocation={Data.coords}
        />
      </div>
      <button
        onClick={() => setButtonPopup(true)}
        className='SmallBtn'
        id='HelpButton'
      >
        help?
      </button>
      <Help trigger={ButtonPopup} setTrigger={setButtonPopup}>
        <p>
          - Click Keen! if you are keen to potentially visit this restaurant
          otherwise click Nope!.
          <br></br>- View this restaurant on the map by clicking the green
          button below the card.
        </p>
      </Help>
      {redirect && <Redirect to={{ pathname: '/Result', state: CardPass }} />}
    </>
  );
}

export default SwipingPage;