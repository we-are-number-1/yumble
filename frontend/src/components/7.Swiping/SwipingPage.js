import React, {useEffect, useState, useContext} from 'react';
import {useHistory, Redirect} from 'react-router-dom';
// import axios from 'axios';

import {SocketContext} from './../../sockets/SocketContext';
import * as SocketEvents from './../../sockets';
import Help from '../Common/Help';
import '../Common/Help.css';
import MapModal from '../Common/MapModal';
import Icon from '../Common/MapsPinpoint';
import '../7.Swiping/SwipingPage.css';
import SwipeCard from '../Common/SwipeCard';

/**
 * @return {*}
 * @param  {*} props
 */
function SwipingPage(props) {
  const history = useHistory();
  const socketContext = useContext(SocketContext);
  const [ButtonPopup, setButtonPopup] = useState(false);
  const [MapPopup, setMapPopup] = useState(false);
  const CardData = props.location.state;
  const [CardPass, setCardPass] = useState(null);

  const RemainingTime = '25';
  const [Data, setData] = useState(CardData[0]); // 0th element
  const [redirect, setRedirect] = useState(false);

  useEffect(() => {
    document.title = 'Yes or No?';
    SocketEvents.endGame(socketContext.socket, goNextPge);
    setCardPass(props.location.state.slice());
    CardData.shift();
  }, []);

  console.log(CardData);
  /**
   */
  const goNextPge = () => {
    console.log('Game has Ended');
    history.replace('/Result');
  };

  /**
  *
  *
  */
  function getNewCard() {
    console.log('CLICK');
    try {
      if (CardData[0] !== undefined) {
        setData(CardData.shift());
      } else {
        setRedirect(true);
      }
    } catch (error) {
      setRedirect(true);
    }
  }

  return (
    <>
      <h1 className='Title'> yumble</h1>
      <h1 className='TimeCounter'> Remaining time: {RemainingTime}s</h1>
      <div className='MakeCentre'>
        <button
          className='YesOrNoButton'
          id='YesButton'
          onClick = {getNewCard}
        >
          Keen!
        </button>
        <button
          className='YesOrNoButton'
          id='NoButton'
          onClick = {getNewCard}
        >
          Nope!
        </button>

        <SwipeCard data = {Data} ></SwipeCard>

        <button
          onClick={() => setMapPopup(true)}
          className='BigBtn'
          id='GoogleMaps_btn'
        >
          View on Google Maps
          <Icon />
        </button>
        <MapModal trigger={MapPopup} setTrigger={setMapPopup}
          restaurantLocation={Data.coords}/>
      </div>
      <button
        onClick={() => setButtonPopup(true)}
        className='SmallBtn'
        id='HelpButton'
      >
        help?
      </button>
      <Help trigger={ButtonPopup} setTrigger={setButtonPopup}>
        <p>
          - Click Keen! if you are keen to potentially visit this restaurant
          otherwise click Nope!.
          <br></br>
          - View this restaurant on the map by clicking the green button
          below the card.
        </p>
      </Help>
      {redirect && <
        Redirect to={{pathname: '/Result', state: CardPass}}/>}
    </>
  );
}

export default SwipingPage;
