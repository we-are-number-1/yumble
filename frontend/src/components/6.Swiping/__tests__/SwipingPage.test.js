import React from 'react';
import renderer from 'react-test-renderer';
import SwipingPage from '../SwipingPage';
import {SocketContextProvider} from '../../../sockets/SocketContext';

jest.mock('./../../../sockets/SocketContext');
jest.mock('./../../../sockets');

test('Swiping page', () => {
  const Data =[{
    name: 'Lonestar',
    coords: 'lat: -36.9685858, lng: 174.8595453',
    images: 'https://c.files.bbci.co.uk/050B/production/_103119210_lazytown2.jpg',
    location: 'Botany',
    cuisine: 'European',
    price: '$$$',
    rating: 4.0,
  }];

  const location={Location};
  location.state = [Data];
  const tree = renderer.create(
      <SocketContextProvider>
        <SwipingPage
          location = {location}/>
      </SocketContextProvider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
