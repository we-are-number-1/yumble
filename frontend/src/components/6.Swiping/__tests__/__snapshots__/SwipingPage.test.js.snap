// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swiping page 1`] = `
Array [
  <h1
    className="Title"
  >
     yumble
  </h1>,
  <h1
    className="TimeCounter"
  >
     Remaining time: 
    s
  </h1>,
  <div
    className="MakeCentre"
  >
    <div
      className="container"
      style={
        Object {
          "marginTop": "2em",
        }
      }
    >
      <div
        className="justify-content-md-center row row-cols-lg-12"
      >
        <div
          className="btnColumn col-md-2 col-6 order-md-1 order-2"
        >
          <button
            className="YesOrNoButton "
            id="NoButton"
            onClick={[Function]}
          >
            Nope!
          </button>
        </div>
        <div
          className="col-lg-6 col-md-8 col-12 order-md-2 order-1"
        >
          <div
            className="CardOverlayText "
          >
            Voted: 
            
            !
          </div>
          <div
            className="ActionableSwipeCard"
          >
            <div
              className="SwipeCard "
              onClick={[Function]}
            >
              <div
                className="CardImage"
              >
                <img
                  className="CardImage"
                  src="https://c.files.bbci.co.uk/050B/production/_103119210_lazytown2.jpg"
                />
              </div>
              <div>
                <p
                  className="PlaceDetails"
                  id="name"
                >
                   
                  Lonestar
                   
                </p>
              </div>
              <p
                className="PlaceDetails"
                id="location"
              >
                 
                Botany
                 
              </p>
              <p
                className="PlaceDetails"
                id="price"
              >
                 
                $$$
                 
              </p>
              <p
                className="PlaceDetails"
                id="rating"
              >
                 
                4.0
                 Rating
              </p>
            </div>
          </div>
        </div>
        <div
          className="btnColumn col-md-2 col-6 order-md-3 order-3"
        >
          <button
            className="YesOrNoButton align-items-center "
            id="YesButton"
            onClick={[Function]}
          >
            Keen!
          </button>
        </div>
      </div>
    </div>
    <button
      className="BigBtn btn btn-warning"
      disabled={false}
      id="GoogleMaps_btn"
      onClick={[Function]}
      type="button"
    >
      View on Google Maps
      <svg
        fill="#ff6056"
        height="30"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 16s6-5.686 6-10A6 6 0 0 0 2 6c0 4.314 6 10 6 10zm0-7a3 3 0 1 1 0-6 3 3 0 0 1 0 6z"
        />
      </svg>
    </button>
    
  </div>,
  <button
    className="btn btn-info btn-lg"
    disabled={false}
    id="HelpButton"
    onClick={[Function]}
    type="button"
  >
    Help
  </button>,
  "",
]
`;
